File: Python/NumberOfIslands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid or not grid[0]:
            return 0
        
        m=len(grid)
        n=len(grid[0])
        
        visited=set()
        queue=collections.deque()
        result=0
        
        for i in range(m):
            for j in range(n):
                if (i,j) in visited:
                    continue
                
                if grid[i][j]=='1':
                    result+=1
                    queue.append((i,j))
                    visited.add((i,j))
                    
                    while queue:
                        x,y=queue.popleft()
                        
                        for (dx,dy) in [(0,1),(1,0),(0,-1),(-1,0)]:
                            X=x+dx
                            Y=y+dy
                            
                            if X>=0 and X<m and Y>=0 and Y<n and grid[X][Y]=='1' and (X,Y) not in visited:
                                queue.append((X,Y))
                                visited.add((X,Y))
        
        return result

File: Java/numberOfIslands.java
class Solution {
    public int numIslands(char[][] grid) {
        if (grid==null || grid.length==0 || grid[0].length==0){
            return 0;
        }
        
        int row=grid.length;
        int col=grid[0].length;
        
        int result=0;
        boolean[][] visited = new boolean[row][col];
        
        for (int i=0;i<row;i++){
            for (int j=0;j<col;j++){
                if(grid[i][j]=='1' && !visited[i][j]){
                    BFS(grid,visited,i,j);
                    result++;
                }
            }
        }
        
        return result;
    }
    
    private void BFS(char[][] grid,boolean[][] visited,int x,int y){
        int[] dx={1,0,0,-1};
        int[] dy={0,1,-1,0};
        
        Queue<Integer> qx = new LinkedList<>();
        Queue<Integer> qy = new LinkedList<>();

        qx.offer(x);
        qy.offer(y);
        visited[x][y] = true;
 
         while (!qx.isEmpty()) {
            int cx = qx.poll();
            int cy = qy.poll();

            for (int i = 0; i < 4; i++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];
                if (0 <= nx && nx < grid.length && 0 <= ny && ny < grid[0].length && !visited[nx][ny] && grid[nx][ny]=='1') {
                    qx.offer(nx);
                    qy.offer(ny);
                    visited[nx][ny] = true;
                }
            }
        }
        
        
    }
}